File: 2305-fair-distribution-of-cookies/2305-fair-distribution-of-cookies.cpp
class Solution {
public:
    
    int ans = INT_MAX;
    
    int distributeCookies(vector<int>& cookies, int k) {
        vector<int> arr(k, 0);
        backtrack(arr, cookies, 0, k);
        
        return ans;
        
    }
    
    void backtrack(vector<int>& arr, vector<int>& cookies,int index, int k){
        if(index == cookies.size()){
            int maxCount = 0;
            for(int i = 0; i < arr.size(); i++){
                if(maxCount < arr[i]){
                    maxCount = arr[i];
                }
            }
            ans = min(ans, maxCount);
            
            return;
        }
        
        for(int i = 0; i < k; i++){
            arr[i] += cookies[index];
            backtrack(arr, cookies, index + 1, k);
            arr[i] -= cookies[index];
        }
    }
};

File: 2305-fair-distribution-of-cookies/2305-fair-distribution-of-cookies.py
class Solution:
    def distributeCookies(self, cookies: List[int], k: int) -> int:
        buckets = [0] * k
        ans = [float("inf")]
        cookies.sort(reverse = True)
        
        def dfs(idx):
            if idx >= len(cookies):
                ans[0] = min(ans[0], max(buckets))
                return
            
            if max(buckets) > ans[0]:
                return
            
            for i in range(k):
                buckets[i] += cookies[idx]
                dfs(idx + 1)
                buckets[i] -= cookies[idx]
        
        dfs(0)
        
        
        return ans[0]